File: apps/api/src/domain/recurring/recurring_rule.ts
import { AggregateRoot, Criteria, Paginate } from "@abejarano/ts-mongodb-criteria";

export type RecurringFrequency = "weekly" | "monthly" | "yearly";

export interface RecurringTemplate {
  type: "income" | "expense" | "transfer";
  amount: number;
  currency: string;
  categoryId?: string;
  fromAccountId?: string;
  toAccountId?: string;
  note?: string;
  tags?: string[];
}

export type RecurringRulePrimitives = {
  id?: string;
  ruleId: string;
  userId: string;
  frequency: RecurringFrequency;
  interval: number;
  startDate: Date;
  endDate?: Date;
  template: RecurringTemplate;
  active: boolean;
  lastRunAt?: Date;
};

export class RecurringRule extends AggregateRoot {
  constructor(private readonly props: RecurringRulePrimitives) {
    super();
  }

  static fromPrimitives(props: RecurringRulePrimitives): RecurringRule {
    return new RecurringRule(props);
  }

  toPrimitives(): RecurringRulePrimitives {
    return { ...this.props };
  }

  get id(): string | undefined {
    return this.props.id;
  }

  get ruleId(): string {
    return this.props.ruleId;
  }

  getId(): string {
    return this.props.id ?? this.props.ruleId;
  }

  get userId(): string {
    return this.props.userId;
  }

  get frequency(): RecurringFrequency {
    return this.props.frequency;
  }

  get interval(): number {
    return this.props.interval;
  }

  get startDate(): Date {
    return this.props.startDate;
  }

  get endDate(): Date | undefined {
    return this.props.endDate;
  }

  get template(): RecurringTemplate {
    return this.props.template;
  }

  get isActive(): boolean {
    return this.props.active;
  }
}

export interface RecurringRuleRepository {
  upsert(rule: RecurringRule): Promise<void>;
  list(criteria: Criteria): Promise<Paginate<RecurringRule>>;
  byId(ruleId: string): Promise<RecurringRule | undefined>;
  remove(ruleId: string): Promise<void>;
}

File: apps/flutter/lib/features/recurring/domain/entities/recurring_rule.dart
class RecurringRule {
  final String id;
  final String userId;
  final String frequency;
  final int interval;
  final DateTime startDate;
  final DateTime? endDate;
  final RecurringTemplate template;
  final bool active;
  final DateTime? lastRunAt;

  RecurringRule({
    required this.id,
    required this.userId,
    required this.frequency,
    required this.interval,
    required this.startDate,
    this.endDate,
    required this.template,
    required this.active,
    this.lastRunAt,
  });

  factory RecurringRule.fromJson(Map<String, dynamic> json) {
    return RecurringRule(
      id: json['recurringRuleId'] ?? json['ruleId'],
      userId: json['userId'],
      frequency: json['frequency'],
      interval: json['interval'],
      startDate: DateTime.parse(json['startDate']),
      endDate: json['endDate'] != null ? DateTime.parse(json['endDate']) : null,
      template: RecurringTemplate.fromJson(json['template']),
      active: json['isActive'] ?? json['active'],
      lastRunAt: json['lastRunAt'] != null
          ? DateTime.parse(json['lastRunAt'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'userId': userId,
      'frequency': frequency,
      'interval': interval,
      'startDate': startDate.toIso8601String(),
      'endDate': endDate?.toIso8601String(),
      'template': template.toJson(),
      'active': active,
      'lastRunAt': lastRunAt?.toIso8601String(),
    };
  }
}

class RecurringTemplate {
  final double amount;
  final String type; // 'income', 'expense', 'transfer'
  final String currency;
  final String? categoryId;
  final String? fromAccountId;
  final String? toAccountId;
  final String? note;
  final List<String>? tags;

  RecurringTemplate({
    required this.amount,
    required this.type,
    required this.currency,
    this.categoryId,
    this.fromAccountId,
    this.toAccountId,
    this.note,
    this.tags,
  });

  factory RecurringTemplate.fromJson(Map<String, dynamic> json) {
    return RecurringTemplate(
      amount: (json['amount'] as num).toDouble(),
      type: json['type'],
      currency: json['currency'],
      categoryId: json['categoryId'],
      fromAccountId: json['fromAccountId'],
      toAccountId: json['toAccountId'],
      note: json['note'],
      tags: (json['tags'] as List?)?.map((e) => e as String).toList(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'amount': amount,
      'type': type,
      'currency': currency,
      'categoryId': categoryId,
      'fromAccountId': fromAccountId,
      'toAccountId': toAccountId,
      'note': note,
      'tags': tags,
    };
  }
}

class RecurringPreviewItem {
  final String recurringRuleId;
  final DateTime date;
  final RecurringTemplate template;
  final String status; // 'new' | 'already_generated'

  RecurringPreviewItem({
    required this.recurringRuleId,
    required this.date,
    required this.template,
    required this.status,
  });

  factory RecurringPreviewItem.fromJson(Map<String, dynamic> json) {
    return RecurringPreviewItem(
      recurringRuleId: json['recurringRuleId'],
      date: DateTime.parse(json['date']),
      template: RecurringTemplate.fromJson(json['template']),
      status: json['status'],
    );
  }
}
